lib/remap-functions/src/join.rs

```diff
@@ -0,0 +1,175 @@
+use remap::prelude::*;
+use std::borrow::Cow;
+
+#[derive(Clone, Copy, Debug)]
+pub struct Join;
+
+impl Function for Join {
+    fn identifier(&self) -> &'static str {
+        "join"
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "value",
+                accepts: |v| matches!(v, Value::Array(_)),
+                required: true,
+            },
+            Parameter {
+                keyword: "separator",
+                accepts: |v| matches!(v, Value::Bytes(_)),
+                required: false,
+            },
+        ]
+    }
+
+    fn compile(&self, mut arguments: ArgumentList) -> Result<Box<dyn Expression>> {
+        let value = arguments.required("value")?.boxed();
+        let separator = arguments.optional("separator").map(Expr::boxed);
+
+        Ok(Box::new(JoinFn { value, separator }))
+    }
+}
+
+#[derive(Clone, Debug)]
+struct JoinFn {
+    value: Box<dyn Expression>,
+    separator: Option<Box<dyn Expression>>,
+}
+
+impl Expression for JoinFn {
+    fn execute(&self, state: &mut state::Program, object: &mut dyn Object) -> Result<Value> {
+        let array = self.value.execute(state, object)?.try_array()?;
+
+        let string_vec = array
+            .iter()
+            .map(|s| s.try_bytes_utf8_lossy().map_err(Into::into))
+            .collect::<Result<Vec<Cow<'_, str>>>>()
+            .map_err(|_| "all array items must be strings")?;
+
+        let separator: String = self
+            .separator
+            .as_ref()
+            .map(|s| {
+                s.execute(state, object)
+                    .and_then(|v| Value::try_bytes(v).map_err(Into::into))
+            })
+            .transpose()?
+            .map(|s| String::from_utf8_lossy(&s).to_string())
+            .unwrap_or_else(|| "".into());
+
+        let joined = string_vec.join(&separator);
+
+        Ok(Value::from(joined))
+    }
+
+    fn type_def(&self, state: &state::Compiler) -> TypeDef {
+        use value::Kind;
+
+        let separator_type = self
+            .separator
+            .as_ref()
+            .map(|separator| separator.type_def(state).fallible_unless(Kind::Bytes));
+
+        self.value
+            .type_def(state)
+            .fallible_unless(Kind::Array)
+            .merge_optional(separator_type)
+            .with_constraint(Kind::Bytes)
+            .with_inner_type(self.value.type_def(state).inner_type_def)
```

Comment:
Is this right? I don't think the output here, `Bytes`, has an inner type.

Reply:
In this case, the inner type is being applied the `value` array, not to the output

Reply:
Is this true? `with_inner_type` seems to set the inner_type of the type being returned by the remap function:

https://github.com/timberio/vector/blob/cf6f56ac088ae525b1e368de64eaf9655487fbd3/lib/remap-lang/src/type_def.rs#L204-L207

Reply:
For example, here it is used to set the inner type for the output of `parse_regex`: https://github.com/timberio/vector/blob/f2529f091786df59ab9d7155c69483f93fec3041/lib/remap-functions/src/parse_regex.rs#L61

Reply:
Ah, yes you are correct. That `with_inner_type` shouldn't be there. That's one other reason it would be nice for `Kind::Array` to be an enum with the inner type. It would then be impossible to specify an inner type for Bytes.

In this case, if you swapped the order of the calls `with_constraint` and `with_inner_type`, the `with_constraint` call would actually wipe out the inner type if the type was a scalar. But since they are in this order the inner_type would linger about..

https://github.com/timberio/vector/blob/master/lib/remap-lang/src/type_def.rs#L194-L199

Reply:
So, I think a good way to look at it is that the purpose of `TypeDef` isn't really to figure out which fields or values have which type. It's basically a fancy machine for providing a yes/no answer to whether the expression is fallible. So the VRL compiler looks at an expression and if `expr.type_def.is_fallible()` (pseudocode) returns `true`, then error handling is required and you get a slap on the wrist if you don't provide it. So whenever you see a chain of methods like this, each step is providing information that informs that final yes/no decision. In this case:

* `type_def` => starts with the default `TypeDef` (with `fallible` set to `false`) and uses the `value` to get as much info as it can (maybe `value` is known, maybe it refers to `.foo` and thus can't be known) and updates `TypeDef` with that info in mind.
* `fallible_unless` => set `fallible` to `true` unless `value` is an array
* `merge_optional` => merge in any fallibility-related information from `separator` if provided
* `with_constraint` => initially, `self.kind` is set to `all` and a constraint sets it to something more restricted; `self.kind` isn't AFAIK used to make fallibility decisions but is required by the compiler to do things like forbidding `join(["foo"], ", ") == 47`.
* `with_inner_type` sets `inner_type_def` and is fully separate from `kind` and thus unrelated to `with_constraint`.

Reply:
Ah, I wrote that without seeing Stephen's comment 😄 I'll go fix that because I can see why the order of operations is important here.

Reply:
I think you just need to move the `fallible_unless_array_has_inner_type` up before `with_constraint` is called. It seems like we need to always have the fallible checks before we add the type def constraints since those modify the type.

Reply:
That is, this patch seems to work:

```diff
diff --git a/lib/remap-functions/src/join.rs b/lib/remap-functions/src/join.rs
index 35131c4a0..08126c7fb 100644
--- a/lib/remap-functions/src/join.rs
+++ b/lib/remap-functions/src/join.rs
@@ -76,9 +76,8 @@ impl Expression for JoinFn {
             .type_def(state)
             .fallible_unless(Kind::Array)
             .merge_optional(separator_type)
-            .with_constraint(Kind::Bytes)
-            .with_inner_type(self.value.type_def(state).inner_type_def)
             .fallible_unless_array_has_inner_type(Kind::Bytes)
+            .with_constraint(Kind::Bytes)
     }
 }
 
@@ -96,7 +95,7 @@ mod test {
             def: TypeDef {
                 fallible: false,
                 kind: Kind::Bytes,
-                inner_type_def: Some(inner_type_def!([ Kind::Bytes ]))
+                ..Default::default()
             },
         }
 
@@ -108,7 +107,7 @@ mod test {
             def: TypeDef {
                 fallible: true,
                 kind: Kind::Bytes,
-                inner_type_def: Some(inner_type_def!([ Kind::Bytes | Kind::Integer ]))
+                ..Default::default()
             },
         }
 
@@ -132,7 +131,7 @@ mod test {
             def: TypeDef {
                 fallible: true,
                 kind: Kind::Bytes,
-                inner_type_def: Some(inner_type_def!([ Kind::Bytes ]))
+                ..Default::default()
             },
         }
```

Reply:
What's happening in this patch, though, is that the `inner_type_def` simply remains `None` throughout (which is the default), and that in turn makes the `fallible_unless_array_has_inner_type` call a no-op.

Reply:
It doesn't remain `None`, it starts at whatever the input value's inner type is. `with_constraint` strips it since we are setting the type to a scalar (`Bytes`). By moving the fallible check up, it is able to access the inner type of the value. You'd run into the same issue had you put `.fallible_unless(Kind::Array)` after you called `with_constraint`. The issue is that `with_constraint` modifies the type.

I think the moral of the story is to just make sure to put the fallible checks before `with_constraint` (or any other methods that mutate the type) is called. That is: call `with_constraint` last.